api/db-cdr/generate-cdr/libproject/genomic_manifests.rb

```diff
@@ -0,0 +1,159 @@
+  :wgs_aw4_prefix => "gs://prod-drc-broad/AW4_wgs_manifest/AoU_DRCB_SEQ_"
+}
+
+
+FILE_ENVIRONMENTS = {
+  "all-of-us-rw-preprod" => {
+    :source_config => PROD_SOURCE_CONFIG,
+  },
+  "all-of-us-rw-prod" => {
+    :source_config => PROD_SOURCE_CONFIG
+  }
+}
+
+FILE_PREFIX_REPLACEMENTS = {
+  :wgs_cram => {
+    # Match/replace the GC filename; this works for both .cram and .cram.crai
+    :matchSourceName => /^[^\/]+\.cram/,
+    :destNameFn => ->(rid) { "wgs_#{rid}.cram" }
+  }
+}
+
+def _aw4_filename_to_datetime(aw4_prefix, f)
+  common = Common.new
+
+  date_match = f.delete_prefix(aw4_prefix).match(/([\d-]+)(_\d+)?.csv/)
+  unless date_match
+    common.warning "AW4 filename does not match expected date format, assuming old: #{f}"
+    return DateTime.new(1970)
+  end
+  date_part = date_match[1]
+  begin
+    return DateTime.strptime(date_part, "%Y-%m-%d-%H-%M-%S")
+  rescue Date::Error
+    common.warning "Found date-part in AW4 filename but cannot parse it, assuming old: #{date_part}"
+    return DateTime.new(1970)
+  end
+end
+
+def _read_research_id_csvs(aw4_prefix, rids)
+  common = Common.new
+
+  ridset = rids.to_set
+
+  # See the CDR playbook for more documentation on AW4s:
+  # https://docs.google.com/document/d/1St6pG_EUFB9oRQUQaOSO7a9UPxPkQ5n4qAVyKF9j9tk/edit#
+  # Read all AW4s into a CSV::Row[]
+  aw4_paths = []
+  if aw4_prefix.start_with?("gs://")
+    aw4_paths = common.capture_stdout(["gsutil", "ls", aw4_prefix +  "*"]).split("\n")
+  else
+    aw4_paths = Dir.glob(aw4_prefix + "*")
+  end
+  if aw4_paths.empty?
+    raise ArgumentError.new("failed to find any matching aw4 manifests @ #{aw4_prefix}")
+  end
+  i = 0
+  matching_aw4_rows = aw4_paths.flat_map do |f|
+    common.status "processed #{i}/#{aw4_paths.length} AW4 manifests" if i % 10 == 0
+    i+=1
+
+    filename_date = _aw4_filename_to_datetime(aw4_prefix, f)
+
+    aw4_str = ""
+    if aw4_prefix.start_with?("gs://")
+      aw4_str = common.capture_stdout(["gsutil", "cat", f])
+    else
+      aw4_str = IO.read(f)
+    end
+    if aw4_str.empty?
+      raise ArgumentError.new("failed to read content from #{f}")
+    end
+    aw4_rows = CSV.parse(aw4_str, headers: true)
+    aw4_rows.each do |row|
+      # Parse the filename for the effective datetime for downstream disambiguation.
+      row["aw4_datetime"] = filename_date
+    end
+
+    aw4_rows.filter do |row|
+      # Filter out rids which weren't requested for publishing.
+      not row.header_row? and ridset.include? row["research_id"]
+    end
+  end
+
+  # Transform into a dict of research ID -> CSV::Row to dedupe
+  by_rid = {}
+  matching_aw4_rows.each do |aw4_row|
+    # Prefer the most recent row for each
+    rid = aw4_row["research_id"]
+    aw4_time = aw4_row["aw4_datetime"]
+    if by_rid.key? rid
+      common.warning "found multiple AW4 rows for #{rid}, using the most recent"
+    end
+    if not by_rid.key? rid or aw4_time < by_rid[rid]["aw4_datetime"]
```

Comment:
looks like this keeps the earlier datetime row, not the most recent

Reply:
🤦  - great catch